sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -1060,16 +1060,33 @@ class GpuRLikeMeta(
     parent: Option[RapidsMeta[_, _, _]],
     rule: DataFromReplacementRule) extends BinaryExprMeta[RLike](expr, conf, parent, rule) {
 
+    private var originalPattern: String = ""
     private var pattern: Option[String] = None
 
+    def optimizeSimplePattern(lhs: Expression, rhs: Expression): GpuExpression = {
+      import RegexOptimizationType._
+      val originalAst = new RegexParser(originalPattern).parse()
+      RegexRewriteUtils.matchSimplePattern(originalAst) match {
+        case StartsWith(s) => GpuStartsWith(lhs, GpuLiteral(s, StringType))
+        case Contains(s) => GpuContains(lhs, GpuLiteral(s, StringType))
+        case NoOptimization => {
+          val patternStr = pattern.getOrElse(throw new IllegalStateException(
+            "Expression has not been tagged with cuDF regex pattern"))
+          GpuRLike(lhs, rhs, patternStr)
+        }
+        case _ => throw new IllegalStateException("Unexpected optimization type")
+      }
+    }
+
     override def tagExprForGpu(): Unit = {
       GpuRegExpUtils.tagForRegExpEnabled(this)
```

Comment:
This method can actually disable regexp on the GPU. This means that these optimizations will never kick in when regexp is disabled. I don't know if that is actually desired. You can look at GpuSplit, where we implemented `transpileToSplittableString` and that codepath is not affected by the regexp enable flag.

Reply:
I think we can keep it as is. From an end user standpoint, if I say that I want to disable regex, and we go ahead and rewrite the query to do something different it might be kind of confusing. But I think that is minor. The main thing I am worried about is if there are situations where we could convert a regular expression into a custom kernel, but the transpiler cannot support it. We are now stuck. That appears to be simple enough to do, and we can do it when we see a need for it.